### PR TITLE
dot-product, inner-product, outer-product, and many more structure functions

### PR DESCRIPTION
**Pull request by sritchie**
Our structure operations have a few incompatibilities with the refman / scmutils implementation that I thought we might get dialed before release.

As of this PR, we have some new generic operations:

- `g/dot-product`, for scalars, differentials, structures, functions and
  row/column matrices
- `g/inner-product` for scalars, structures, functions and row/column matrices
- `g/outer-product` for functions, structures of length 3 and matrices, between
  a row and a column only
- `g/cross-product` now works for row/column matrices in addition to structures
  (and functions that accept these)

Structures got a lot of love too:

- Structures and matrices both gain the ability to do native `get-in`,
  `assoc-in` and `empty`. These work as expected, like a potentially nested
  vector. ([#193](https://github.com/littleredcomputer/sicmutils/pull/193))

- `matrix.cljc` gains `up->row-matrix`, `up->column-matrix`, `row-matrix->up`,
  `column-matrix->up`
  ([#193](https://github.com/littleredcomputer/sicmutils/pull/193))

- `structure.cljc` gains many features in
  ([#193](https://github.com/littleredcomputer/sicmutils/pull/193)):

  - `kronecker` and `basis-unit` for generating potentially infinite basis
    vectors
  - the ability to conj new items onto a structure: `(conj (up 1 2) 3) => (up 1
    2 3)`
  - The structure-preserving `map-chain` takes a 2-arg function and presents it
    with each element of a deeply nested structure, along with a vector of its
    "chain", the path into its location. The fn's return becomes the new item at
    that location.
  - `structure->prototype` generates a same-shape structure as its argument,
    with symbolic entries that display their location.
  - `typical-object` returns a structure of the same shape and orientation as
    `s`, generated by substituting gensymmed symbols in for each entry.
  - `compatible-zero` returns a structure compatible for multiplication with `s`
    down to 0.
  - `transpose-outer` returns a new structure with the same orientation as the
    first element of `s`, filled with elements of the same orientation as `s`.
    Each element is generating by taking the first element of each entry in `s`,
    the the second, etc... In that sense this is similar to a traditional matrix
    transpose.
  - `dot-product` takes the dot product of two structures. They must be the same
    top-level orientation and dimension; beyond that, their entries are
    pairwise-multiplied and summed.
  - `inner-product` is the same, but the left structure is conjugated first.
  - `outer-product` now works multiple levels deep.
  - `vector-outer-product` and `vector-inner-product` are similar, but only
    enforce the top-level length; all internal structures are NOT flattened and
    must be compatible for `g/*`.
  - `compatible-for-contraction?` now searches recursively down into a
    structure; previously it only checked the top level.
  - The new `*allow-incompatible-multiplication*` dynamic variable is set to
    `true` by default. Set it false to force a setting where, when you multiply
    structures, they must be:
    - opposite orientation
    - every element of the right entry must be compatible for contraction with
      the left
  - structure multiplication with scalars, etc now respects ordering, just in
    case any multiplication is not commutative.

### CODE REVIEW

File: test/sicmutils/examples/double_pendulum_test.cljc [hunk 1]
```diff
@@ -85,7 +85,7 @@
                 "  var _000e = Math.cos(_000a);\n"
                 "  var _0010 = Math.sin(_000a);\n"
                 "  var _0011 = Math.pow(_0010, 2);\n"
-                "  return [1, [thetadot, phidot], [(- l1 * m2 * _0006 * _0010 * _000e - l2 * m2 * _0005 * _0010 + g * m2 * _000e * _0008 - g * m1 * _0009 - g * m2 * _0009) / (l1 * m2 * _0011 + l1 * m1), (l2 * m2 * _0005 * _0010 * _000e + l1 * m1 * _0006 * _0010 + l1 * m2 * _0006 * _0010 + g * m1 * _0009 * _000e + g * m2 * _0009 * _000e - g * m1 * _0008 - g * m2 * _0008) / (l2 * m2 * _0011 + l2 * m1)]];\n"
```
Comment (sritchie) [inline test/sicmutils/examples/double_pendulum_test.cljc:88]: that structure change caused some of these tests to switch variable order.

File: test/sicmutils/generators.cljc [hunk 2]
```diff
@@ -104,12 +105,75 @@
               d)]
       (r/rationalize n d))))
 
+;; ## Structure Generators
+
+(defn- recursive
```
Comment (sritchie) [inline test/sicmutils/generators.cljc:110]: OH, you are going to like this. Generative testing with nested structures!

File: test/sicmutils/generic_test.cljc [hunk 3]
```diff
@@ -148,13 +148,11 @@
...
 (deftest generic-times
   (is (= 1 (g/*)) "No args returns the multiplicative identity.")
-  (checking "g/*"
-            100
-            [x gen/any]
-            (is (= x (g/* x)) "single arg returns itself.")
-            (is (= (if (v/one? x) 1 x)
-                   (g/* x 1)) "First unity gets returned.")
-            (is (= x (g/* 1 x)) "Anything times a 1 returns itself.")))
+  (checking "g/*" 100 [x gen/any]
```
Comment (sritchie) [inline test/sicmutils/generic_test.cljc:151]: `v/=` replaces `=`, for the rarish case of `(= 0 0.0)` turning up false.

File: test/sicmutils/structure_test.cljc [hunk 4]
```diff
@@ -20,8 +20,13 @@
 (ns sicmutils.structure-test
   (:refer-clojure :exclude [+ - * /])
   (:require [clojure.test :refer [is deftest testing]]
+            [clojure.test.check.generators :as gen]
```
Comment (sritchie) [inline test/sicmutils/structure_test.cljc:23]: I added lots of tests here, generative style, and re-arranged what we had already to go in roughly the order we have in the namespace.

**Comment by sritchie** [comment]
@littleredcomputer , I ended up spending another day fleshing out the `structure.cljc` namespace with ~everything that GJS has in scmutils (no abstract structures yet).

This uncovered a few bugs with the way we were handling deeply nested structures. It didn't matter as much (except the double pendulum equations of motion had an issue that you'll see), but the behavior of the `outer-product` and `inner-product` that we had named in `structure.cljc` did not match what we want to see out of the tests.

I think this brings us up to as much parity with the refman as I could hope for.

The PR is pretty big because of a ton of tests. I am happy to split this up - most of the reading is in `structure.cljc` and its tests, and that is all fairly smooth to follow.

Would love feedback here, since I know you've been deep in the Structure Mines and have thoughts here!

File: test/sicmutils/examples/central_potential_test.cljc [hunk 5]
```diff
@@ -130,21 +132,15 @@
...
-                          (up (up 0
-                                  (* (/ 1 2) m (((expt D 2) x) t)))
-                              (up (* (/ 1 2) m (((expt D 2) x) t))
-                                  (* m (((expt D 2) y) t)))))
-                    (down (up (up 0 0)
-                              (up 0 0))
-                          (up (up 0 0)
-                              (up 0 0))))
+
+      (is (= '(down (down (* m (((expt D 2) x) t))
```
[the remark this answers is not in the document]
Comment (littleredcomputer) [reply test/sicmutils/examples/central_potential_test.cljc:136]: which change?

[the remark this answers is not in the document]
Comment (sritchie) [reply test/sicmutils/examples/central_potential_test.cljc:136]: Our original `g/square` implementation only went one level deep; the `scmutils` one is able to recursively merge structures by using the `structure` implementation of `dot-product`, and not the same function as `*` that we were calling.

scheme:

```
1 ]=> (square (up (up 1) (up 2)))
#| 5 |#
```

sicmutils before the change, using the old `inner-product` impl:

```clojure
sicmutils.structure>  (g/square (up (up 1) (up 2)))
#object[sicmutils.structure.Structure 0x1caa9d7d "(up (up 5))"]
```

after the change, using the new `dot-product`:

```clojure
sicmutils.structure>  (g/square (up (up 1) (up 2)))
5
```

File: test/sicmutils/generators.cljc (hunk 2 above)
Comment (littleredcomputer) [reply test/sicmutils/generators.cljc:110]: Yup. This is brilliant!

File: test/sicmutils/matrix_test.cljc [hunk 6]
```diff
@@ -373,6 +392,44 @@
...
+              (is (ish? (m/up->row-matrix
+                         (g/cross-product s s))
+                        (g/cross-product row row)))))
+
+  (checking "outer-product" 100 [s (sg/up1 sg/complex 2 10)]
+            (let [col (m/up->column-matrix s)
+                  row (m/up->row-matrix s)]
+              (is (= (g/outer-product s s)
+                     (g/outer-product col row))))))
+
```
Comment (littleredcomputer) [inline test/sicmutils/matrix_test.cljc:432]: You could have (unless I overlooked it) a test that asserts that `(* (transpose v) v)` is _square_ for any structure v

**Review by littleredcomputer — APPROVED** [review]

File: src/sicmutils/matrix.cljc [hunk 7]
```diff
@@ -425,6 +443,34 @@
...
+
+(defn up->row-matrix
+  "Returns a row matrix with the contents of the supplied `up` structure.
+  Errors if any other type is provided."
+  [v]
+  {:pre [(s/up? v)]}
+  (by-rows (seq v)))
+
+(defn row-matrix->up
+  "Extracts the single row from the supplied column matrix. Errors if some other
```
[the remark this answers is not in the document]
Comment (sritchie) [reply src/sicmutils/matrix.cljc:468]: fixed in a future one!

File: test/sicmutils/matrix_test.cljc (hunk 6 above)
Comment (sritchie) [reply test/sicmutils/matrix_test.cljc:432]: is that true? I think any structure now will collapse to a number...

That IS true for `transpose-outer` with 2-tensors. Good test, will add.

File: src/sicmutils/structure.cljc [hunk 8]
```diff
@@ -345,104 +472,67 @@
...
+  - a structure `s`
+
+  and returns a new structure of identical shape, with symbolic entries like
+  `'x:0:1` that show their access chain."
+  [name s]
+  (mapr (fn [chain]
+          (let [path (join ":" chain)]
+            (symbol
+             (str name path))))
+        (structure->access-chains s)))
```
[the remark this answers is not in the document]
Comment (sritchie) [reply src/sicmutils/structure.cljc:526]: Yes, this is awesome. I'm going to do this in a separate PR. I can do it be returning a sequence of `[::up ::down ::down]` etc attached as metadata to the chain... then anyone can get access to this info if they want, and we can use it here.

That will, as you say, give us the property that this structure will match our literal structures AND our literal matrix notation.

File: src/sicmutils/structure.cljc [hunk 9]
```diff
@@ -209,13 +240,118 @@
...
+               (= (.-v this)
+                  (.-v s))))
+
+        (= (.-orientation this) ::up)
+        (cond (vector? that)   (= (.-v this) that)
+              (seqable? that) (= (seq this) (seq that))
+              :else false)
+        :else false))
+
+(defn structure?
```
[the remark this answers is not in the document]
Comment (sritchie) [reply src/sicmutils/structure.cljc:321]: Good idea. I'll do that in the same followup as the other suggestion.

File: src/sicmutils/structure.cljc [hunk 10]
```diff
@@ -42,14 +42,33 @@
...
+  [i j]
+  (if (== i j) 1 0))
+
+(defn basis-unit
+  "Returns a basis vector of `n` 0s, with `1` in the `i`th position.
+
+  If `n` is not supplied returns an infinite sequence."
+  ([i] (map (partial kronecker i)
+            (range)))
+  ([n i] (into [] (take n) (basis-unit i))))
```
[the remark this answers is not in the document]
Comment (sritchie) [reply src/sicmutils/structure.cljc:66]: good idea, I for some reason thought this was the defensive way to go, but it looks like any place this is used makes a vector if it needs to.

File: src/sicmutils/examples/central_potential.cljc [hunk 11]
```diff
@@ -24,33 +24,36 @@
...
-              (map (fn [[[m1 p1] [m2 p2]]]
-                     (/ (* m1 m2) (sqrt (square (- p1 p2)))))
-                   mass-position-pairs)))))
+    (let [mass-position-pairs (->> (partition 2 x)
+                                   (map (fn [m [x y]] [m (up x y)])
+                                        masses)
+                                   (pairs))]
+      (reduce - 0 (map (fn [[[m1 p1] [m2 p2]]]
+                         (/ (* m1 m2)
+                            (sqrt (square (- p1 p2)))))
```
[the remark this answers is not in the document]
Comment (sritchie) [reply src/sicmutils/examples/central_potential.cljc:44]: will do in my followup!

**Pull request merged**